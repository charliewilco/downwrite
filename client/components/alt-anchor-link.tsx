import * as React from "react";
import classNames from "../utils/classnames";

interface AltAnchorProps<T> extends React.AnchorHTMLAttributes<T> {
  space?: string;
}

export default function AltAnchorLink({
  space,
  className,
  ...props
}: AltAnchorProps<any>): JSX.Element {
  const cx = classNames("AltLink", className);
  return <a {...props} className={cx} />;
}
