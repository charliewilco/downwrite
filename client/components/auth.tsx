import * as React from "react";
import Cookies from "universal-cookie";
import Router from "next/router";
import * as jwt from "jwt-decode";
import addDays from "date-fns/add_days";
import { NextComponentClass } from "next";
import { __IS_TEST__ } from "../utils/dev";

// NOTE:
// This component should passdown the state of authed from withAuthCheck() HOC
// There is only one single point of state that needs to be rendered

// One other pattern we could consider is passing down user and token as state
// and login and logout functions from authed. this would allow an initial check
// of the cookie on a refresh and as the user is logged in have an updated source of the token
// this would solve that single point of state to be updated.
// We would pass down signIn and signOut to <Login /> & <Register />

/*
	<Auth>
		{(authed, login, logout, token, user) => <App {...args} />}
	</Auth>
*/

// state needs to evaluate the existence of
// token + decode the content of the token

const cookie = new Cookies();
const COOKIE_EXPIRATION = 180;
const cookieOptions = {
  path: "/",
  expires: addDays(Date.now(), COOKIE_EXPIRATION)
};

interface IAuthProps {
  children: React.ReactNode;
  token: string;
  authed: boolean;
}

interface IAuthState {
  token: string;
  name?: string;
  authed: boolean;
}

export interface IAuthActions {
  signIn: (authed: boolean, token: string) => void;
  signOut: () => void;
}

export interface IAuthContext extends IAuthState, IAuthActions {}

interface IToken {
  name: string;
  user: string;
}

const EMPTY_USER: IToken = {
  user: null,
  name: null
};

export const AuthContext = React.createContext({} as IAuthContext);

// Should be able to just request user details from another call
export default class AuthMegaProvider extends React.Component<
  IAuthProps,
  IAuthState
> {
  constructor(props: IAuthProps) {
    super(props);

    let token = this.props.token || cookie.get("DW_TOKEN");

    let __TOKEN_EXISTS__: boolean = token !== undefined && token !== "undefined";
    const { name } = __TOKEN_EXISTS__ ? jwt<IToken>(token) : EMPTY_USER;

    this.state = {
      token,
      authed: __TOKEN_EXISTS__,
      name: name || null
    };
  }

  private signIn = (authed: boolean, token: string): void => {
    const { name } = jwt(token);
    this.setState({ authed, token, name });
    cookie.set("DW_TOKEN", token, cookieOptions);
  };

  private signOut = (): void => {
    this.setState({
      authed: false,
      token: undefined,
      name: undefined
    });
  };

  public componentDidUpdate(prevProps: IAuthProps, prevState: IAuthState): void {
    const { authed } = this.state;
    if (prevState.authed !== authed) {
      Router.push({ pathname: authed ? "/" : "/login" });

      if (!authed) {
        cookie.remove("DW_TOKEN", cookieOptions);
      }
    }
  }

  public render(): JSX.Element {
    return (
      <AuthContext.Provider
        value={{
          ...this.state,
          signIn: this.signIn,
          signOut: this.signOut
        }}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AuthConsumer = AuthContext.Consumer;

// TODO: Remove
// Component should be NextStatelessComponent type
export const withAuth = (Component: NextComponentClass<any>) => {
  return class extends React.Component<any, any> {
    public static displayName = `withAuth(${Component.displayName ||
      Component.name})`;

    public static getInitialProps = Component.getInitialProps;

    public render(): JSX.Element {
      return (
        <AuthContext.Consumer>
          {(auth: IAuthActions) => <Component {...this.props} {...auth} />}
        </AuthContext.Consumer>
      );
    }
  };
};
