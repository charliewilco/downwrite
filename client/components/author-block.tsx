import * as React from "react";
import Link from "next/link";
import Avatar from "./avatar";

interface IAuthorProps {
  authed: boolean;
  name: string;
  colors: string[];
}

function TeaserCopy(): JSX.Element {
  return <>You can write and share on Downwrite, you can sign up or log in </>;
}

export default function Author(props: IAuthorProps): JSX.Element {
  return (
    <aside className="AuthorBlock">
      <header className="AuthorBlockHeader">
        <Avatar colors={props.colors} />
        <h6 className="AuthorBlockTitle">Post from {props.name}</h6>
      </header>
      {!props.authed && (
        <>
          <hr />
          <p className="AuthorBlockContent">
            <TeaserCopy />
            <Link prefetch href="/login">
              <a>here</a>
            </Link>
          </p>
        </>
      )}
    </aside>
  );
}
