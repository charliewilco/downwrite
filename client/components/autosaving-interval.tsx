import * as React from "react";
import delay from "delay";

interface IAutosavingProps {
  onUpdate: (x?: any) => void;
  delay?: number;
  duration?: number;
}

const AutosavingInterval: React.FC<IAutosavingProps> = function(props) {
  const [autosaving, setAutoSaving] = React.useState<boolean>(false);

  let interval: NodeJS.Timeout;

  React.useEffect(() => {
    interval = setInterval(async () => {
      if (props.onUpdate) {
        setAutoSaving(true);
        await delay(props.delay || 3000);
        props.onUpdate();
        setAutoSaving(false);
      }
    }, props.duration || 5000);

    return function cleanup() {
      clearInterval(interval);
    };
  }, []);

  if (autosaving) {
    return <>{props.children}</>;
  }

  return null;
};

export default AutosavingInterval;
