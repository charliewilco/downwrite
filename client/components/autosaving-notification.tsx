import React from "react";
import Toast from "./toast";

export default () => <Toast>Autosaving</Toast>;
