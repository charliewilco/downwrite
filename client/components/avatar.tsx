import * as React from "react";
import classNames from "../utils/classnames";

export type GradientColors = string[];

export const AvatarColors: GradientColors = ["#FEB692", "#EA5455"];

export interface IPointedGradientColors {
  a: string;
  b: string;
}

export const gradientPoints = (colors: GradientColors = AvatarColors) => ({
  a: colors[0],
  b: colors[1]
});

export interface IAvatarCircleProps {
  colors: IPointedGradientColors;
  centered?: boolean;
  size?: number;
}

interface IAvatarProps {
  colors: GradientColors;
  size?: number;
  centered?: boolean;
  className?: string;
}

export default function Avatar(props: IAvatarProps): JSX.Element {
  const cx = classNames(
    "AvatarCircle",
    props.className,
    props.centered && "AvatarCircle--centered"
  );

  const colors = gradientPoints(props.colors);

  const styles = {
    "--size": props.size || 48,
    "--colors-a": colors.a,
    "--colors-b": colors.b
  } as React.CSSProperties;

  return (
    <>
      <div role="image" className={cx} style={styles} />
    </>
  );
}
