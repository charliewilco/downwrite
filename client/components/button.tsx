import styled from "styled-components";
import * as DefaultStyles from "../utils/defaultStyles";

export const Button = styled.button`
  background-color: ${DefaultStyles.colors.yellow700};
  border-radius: 0.25rem;
  border: 0;
  box-shadow: 0 0 4px rgba(0, 0, 0, 0.037), 0 4px 8px rgba(0, 0, 0, 0.07);
  box-sizing: inherit;
  color: #282828;
  display: block;
  font-family: inherit;
  font-size: 14px;
  font-weight: 700;
  line-height: inherit;
  padding: 0.25rem 1.125rem;
  transition: background-color 250ms ease-in-out;
  cursor: pointer;

  &:hover {
    background-color: ${DefaultStyles.colors.yellow500};
  }

  &[disabled] {
    filter: grayscale(100%);
  }
`;

export const AltButton = styled.button`
  margin-right: 1rem;
  color: ${DefaultStyles.colors.blue700};
  background: none;
  font-family: inherit;
  font-size: 100%;
  border: 0px;
  font-weight: 700;
  box-sizing: inherit;
`;
