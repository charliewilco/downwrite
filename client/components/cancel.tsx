import styled from "styled-components";
import * as DefaultStyles from "../utils/defaultStyles";

export default styled.button`
  margin-right: 1rem;
  color: ${DefaultStyles.colors.blue700};
  background: none;
  font-family: inherit;
  font-size: 100%;
  border: 0px;
  font-weight: 700;
  box-sizing: inherit;
`;
