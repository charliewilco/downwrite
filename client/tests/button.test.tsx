import * as React from "react";
import "jest-dom/extend-expect";
import { Button } from "../components/button";
import { render, fireEvent } from "react-testing-library";

const onClickHandler = jest.fn();
const { getByText } = render(
  <Button data-testid="TESTING_BUTTON" onClick={onClickHandler}>
    Button
  </Button>
);

describe("<Button />", () => {
  it("Fires onClick", () => {
    fireEvent.click(getByText("Button"));
    expect(onClickHandler).toHaveBeenCalled();
  });
});
