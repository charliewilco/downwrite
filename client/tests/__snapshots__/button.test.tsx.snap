// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
.c0 {
  background-color: #ffb600;
  border-radius: 0.25rem;
  border: 0;
  box-shadow: 0 0 4px rgba(0,0,0,0.037),0 4px 8px rgba(0,0,0,0.07);
  box-sizing: inherit;
  color: #282828;
  display: block;
  font-family: inherit;
  font-size: 14px;
  font-weight: 700;
  line-height: inherit;
  padding: 0.25rem 1.125rem;
  -webkit-transition: background-color 250ms ease-in-out;
  transition: background-color 250ms ease-in-out;
  cursor: pointer;
}

.c0:hover {
  background-color: #ffc200;
}

.c0[disabled] {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
}

<button
  class="c0"
  data-testid="TESTING_BUTTON"
>
  Button
</button>
`;
